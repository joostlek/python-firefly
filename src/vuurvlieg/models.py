"""Models for Firefly."""

from __future__ import annotations


from dataclasses import dataclass, field
from datetime import date
from enum import StrEnum

from mashumaro import DataClassDictMixin, field_options
from mashumaro.mixins.orjson import DataClassORJSONMixin
from mashumaro.types import SerializationStrategy
from mashumaro.config import BaseConfig


@dataclass
class AboutResponse(DataClassORJSONMixin):
    """AboutResponse model."""

    data: About

@dataclass
class About(DataClassORJSONMixin):
    """About model."""

    version: str
    api_version: str
    php_version: str
    os: str
    driver: str
