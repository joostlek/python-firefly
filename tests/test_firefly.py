"""Asynchronous Python client for Firefly."""

from __future__ import annotations

from typing import TYPE_CHECKING

import aiohttp
from aioresponses import aioresponses
import pytest

from vuurvlieg import Vuurvlieg, FireflyError
from tests import load_fixture

from .const import FIREFLY_URL

if TYPE_CHECKING:
    from syrupy import SnapshotAssertion


async def test_putting_in_own_session(
    responses: aioresponses,
) -> None:
    """Test putting in own session."""
    responses.get(
        f"{FIREFLY_URL}/api/v1/about",
        status=200,
        body=load_fixture("about.json"),
    )
    async with aiohttp.ClientSession() as session:
        vuurvlieg = Vuurvlieg(session=session, api_host="https://demo.firefly.io")
        await vuurvlieg.get_about()
        assert vuurvlieg.session is not None
        assert not vuurvlieg.session.closed
        await vuurvlieg.close()
        assert not vuurvlieg.session.closed


async def test_creating_own_session(
    responses: aioresponses,
) -> None:
    """Test creating own session."""
    responses.get(
        f"{FIREFLY_URL}/api/v1/about",
        status=200,
        body=load_fixture("about.json"),
    )
    vuurvlieg = Vuurvlieg(api_host="https://demo.firefly.io", token="XXX")
    await vuurvlieg.get_about()
    assert vuurvlieg.session is not None
    assert not vuurvlieg.session.closed
    await vuurvlieg.close()
    assert vuurvlieg.session.closed


async def test_unexpected_server_response(
    responses: aioresponses,
    firefly_client: Vuurvlieg,
) -> None:
    """Test handling unexpected response."""
    responses.get(
        f"{FIREFLY_URL}/api/v1/about",
        status=200,
        headers={"Content-Type": "plain/text"},
        body="Yes",
    )
    with pytest.raises(FireflyError):
        assert await firefly_client.get_about()


async def test_about(
    responses: aioresponses,
    firefly_client: Vuurvlieg,
    snapshot: SnapshotAssertion,
) -> None:
    """Test retrieving about."""
    responses.get(
        f"{FIREFLY_URL}/api/v1/about",
        status=200,
        body=load_fixture("about.json"),
    )
    assert await firefly_client.get_about() == snapshot
