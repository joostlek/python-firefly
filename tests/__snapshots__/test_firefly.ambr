# serializer version: 1
# name: test_about
  dict({
    'api_version': '2.1.0',
    'driver': 'sqlite',
    'os': 'Linux',
    'php_version': '8.3.9',
    'version': '6.1.19',
  })
# ---
